Dockerfile

```diff
@@ -1,66 +1,453 @@
+        liblzma-dev lzma lzma-dev tk-dev uuid-dev libmpdec-dev \
+        libnuma-dev
+
+    # Cleanup
+    rm -rf /var/tmp/* \
+        && rm -rf /tmp/* \
+        && rm -rf /var/cache/apt
+EOF
+
+#
+# Stage FlashInfer Build
+#
+# Example build command:
+#   docker build --tag=gpustack/gpustack:cuda-flashinfer-build --file=Dockerfile --target flashinfer-build --progress=plain .
+#
+
+FROM base-build AS flashinfer-build
+
+ARG FLASHINFER_VERSION
+ARG FLASHINFER_TORCH_CUDA_ARCH
+
+ENV FLASHINFER_VERSION=${FLASHINFER_VERSION}
+
+## Install FlashInfer Build Dependencies
+
+RUN <<EOF
+    # FlashInfer
+
+    # Install dependencies,
+    # extract https://github.com/flashinfer-ai/flashinfer/blob/main/scripts/run-ci-build-wheel.sh.
+    cat <<EOT >/tmp/requirements.txt
+torch==2.7.0
+ninja
+numpy
+packaging
+wheel
+build
+EOT
+    pip install --disable-pip-version-check --no-cache-dir --root-user-action ignore -r /tmp/requirements.txt
+
+    # Version
+    IFS="." read -r CUDA_MAJOR CUDA_MINOR CUDA_PATCH <<< "${CUDA_VERSION}"
+    IFS="." read -r TORCH_MAJOR TORCH_MINOR TORCH_PATCH <<< "$(pip freeze | grep torch== | head -n 1 | cut -d'=' -f3 | cut -d'.' -f1-3)"
+
+    # Download
+    git clone --depth 1 --recursive --shallow-submodules \
+        --branch v${FLASHINFER_VERSION} \
+        https://github.com/flashinfer-ai/flashinfer.git /tmp/flashinfer
+
+    # Build
+    TORCH_CUDA_ARCH_LIST="${FLASHINFER_TORCH_CUDA_ARCH}"
+    if [[ -z "${TORCH_CUDA_ARCH_LIST}" ]]; then
+        if [[ "${CUDA_VERSION}" == 11.* ]]; then
+            TORCH_CUDA_ARCH_LIST="7.5 8.0 8.9"
+        elif [[ "${CUDA_VERSION}" == 12.[0-7]* ]]; then
+            TORCH_CUDA_ARCH_LIST="7.5 8.0 8.9 9.0a"
+        else
+            TORCH_CUDA_ARCH_LIST="7.5 8.0 8.9 9.0a 10.0a 12.0"
+        fi
+    fi
+    pushd /tmp/flashinfer
+    TORCH_CUDA_ARCH_LIST="${TORCH_CUDA_ARCH_LIST}" \
+        python -v -m flashinfer.aot
+    FLASHINFER_LOCAL_VERSION="cu${CUDA_MAJOR}${CUDA_MINOR}torch${TORCH_MAJOR}.${TORCH_MINOR}" \
+    TORCH_CUDA_ARCH_LIST="${TORCH_CUDA_ARCH_LIST}" \
+        python -v -m build --no-isolation --wheel
+    popd
+
+    # Archive
+    mkdir -p /workspace/flashinfer \
+        && mv /tmp/flashinfer/dist /workspace/flashinfer/
+
+    # Cleanup
+    rm -rf /var/tmp/* \
+        && rm -rf /tmp/* \
+        && pip cache purge
+EOF
+
+#
+# Stage GPUStack
+#
+# Example build command:
+#   docker build --tag=gpustack/gpustack:cuda --file=Dockerfile --progress=plain .
+#
+
+FROM base AS gpustack
+
+## Install GPUStack
+
+RUN --mount=type=bind,target=/workspace/gpustack,rw <<EOF
+    # GPUStack
+
+    # Build GPUStack
+    export PATH="${HOME}/.local/bin:${PATH}"
+    cd /workspace/gpustack \
+        && git config --global --add safe.directory /workspace/gpustack \
+        && make build
+
+    # Pre process
+    # - Create virtual environment to place gpustack
+    python -m venv --system-site-packages ${PIPX_LOCAL_VENVS}/gpustack
```

Comment:
Can we install gpustack in the default python environment? It would be simpler to install any other missing packages when needed for certain models.